### PR TITLE
[influxdb] Add support for InfluxDB 3.0

### PR DESCRIPTION
**Pull request by cedricboon**
# Description

**Classification:** Novel Addition — adds InfluxDB 3.0 as a third selectable `version` for the InfluxDB persistence service, alongside the existing 1.X and 2.X support.

InfluxDB 3 has a different model from 1.x/2.x: no org/bucket, just a single database addressed with a bearer token, and no retention-policy concept. The official `influxdb3-java` client wraps Apache Arrow Flight, gRPC and Netty (including native boringssl bindings) as transitive dependencies, which adds risks to package correctly in this project's OSGi/bnd runtime. Instead, this implementation talks to the InfluxDB 3 HTTP API directly: writes use line protocol via `POST /api/v3/write_lp`, queries use InfluxQL via `GET /api/v3/query_influxql`, and health checks use `GET /ping`. This reuses OkHttp and Gson, both already direct dependencies of this bundle for InfluxDB 1.x/2.x, so no new dependencies are introduced. InfluxQL query generation reuses the same `org.influxdb.querybuilder` builder as the existing V1 implementation, without the retention-policy-qualified table name, which V3 databases don't have.

For end users:
- `version` now also accepts `V3`.
- For `V3`, `token` is required (no user/password fallback), `db` is the database name, and `retentionPolicy` is unused, since V3 has no retention-policy concept.
- `remove()` (delete-by-predicate) and `getItemInfo` (stored item counts) are not supported for V3, same as for V1, since InfluxDB 3 Core/OSS has no general delete-by-predicate API and no efficient InfluxQL equivalent to V2's Flux-based cross-measurement grouping query.
- The README has been updated accordingly.

# Testing

Unit tests were extended to cover V3: `InfluxFilterCriteriaQueryCreatorImplTest` for InfluxQL query generation, and `InfluxDBPersistenceServiceTest` for configuration validation. `mvn clean test` passes for the bundle.

This was also tested end-to-end against an InfluxDB 3 Enterprise 3.11.2 instance.

A pre-built JAR for testing:
https://raw.githubusercontent.com/cedricboon/openhab-addons/refs/heads/influxdb-v3-support-compiled/bundles/org.openhab.persistence.influxdb/target/org.openhab.persistence.influxdb-5.3.0-SNAPSHOT.jar

### CODE REVIEW

**Comment by lsiepel** [comment]
Wouldn;t this be a good oppertunity to remove V1 or call it deprecated and mark it for removal in openHAB 6

**Comment by cedricboon** [comment]
Not sure about this, I was under the impression that v1 is still used a lot, given the big difference between v1 and v2, and again between v2 and v3

**Comment by lsiepel** [comment]
> Not sure about this, I was under the impression that v1 is still used a lot, given the big difference between v1 and v2, and again between v2 and v3

V1 => V3 seems to be more similar. We can also do this later.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBVersion.java [hunk 1]
```diff
@@ -23,5 +23,6 @@
 public enum InfluxDBVersion {
     V1,
     V2,
+    V3,
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBVersion.java:26]: Adding the enum value does not expose V3 in Main UI. `OH-INF/config/config.xml` still offers only `V1` and `V2`, and its token/database/retention help text plus the generated English i18n entries still describe only V1/V2. Consequently, UI users cannot select the new backend despite the README saying the service is configurable there. Please add the V3 option and update the English configuration metadata; locale-specific translations should remain untouched.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBConfiguration.java [hunk 2]
```diff
@@ -82,6 +82,8 @@ public boolean isValid() {
             hasCredentials = !user.isBlank() && !password.isBlank();
         } else if (version == InfluxDBVersion.V2) {
             hasCredentials = !token.isBlank() || (!user.isBlank() && !password.isBlank());
+        } else if (version == InfluxDBVersion.V3) {
+            hasCredentials = !token.isBlank();
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBConfiguration.java:86]: V3 is documented as having no retention-policy setting, but the shared `valid` expression below still requires `retentionPolicy` to be nonblank. A file-based V3 configuration that explicitly leaves this unused property empty is rejected. Make the retention-policy requirement version-specific and add a V3 test with an empty value.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+    @Override
+    public boolean write(List<InfluxPoint> influxPoints) {
+        final OkHttpClient currentClient = this.client;
+        if (currentClient == null) {
+            return false;
+        }
+
+        String lineProtocolBody = influxPoints.stream().map(this::convertPointToLineProtocol)
+                .filter(Optional::isPresent).map(Optional::get).reduce((a, b) -> a + "\n" + b).orElse("");
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:133]: This reduction copies the accumulated string for every point, making batch construction quadratic. Because the shared queue is unbounded and drains the full outage backlog at once, recovery after a long outage can consume excessive CPU and memory. Use a linear join, such as `Optional::stream` followed by `Collectors.joining("\n")`.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        if (lineProtocolBody.isEmpty()) {
+            return true;
+        }
+
+        HttpUrl url = HttpUrl.get(configuration.getUrl()).newBuilder().addPathSegments("api/v3/write_lp")
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:138]: The URL and request are built before the guarded block. `HttpUrl.get(...)` throws `IllegalArgumentException` for a malformed configured URL; `connect()` catches that during ping but leaves `client` non-null, so the first queued write can let this exception escape `commit()` and permanently cancel the fixed-delay job. Validate the URL when loading configuration or include request construction in the guarded path here and in `query()`.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                .addQueryParameter("db", configuration.getDatabaseName()).addQueryParameter("precision", "ms").build();
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:139]: [The V3 endpoint defaults to partial writes](https://docs.influxdata.com/influxdb3/core/write-data/http-api/v3-write-lp/#partial-writes): it can persist valid lines and still return HTTP 400 for rejected lines. This method maps that response to `false`, after which `commit()` requeues the entire batch and disconnects, so already accepted points are retried and the batch result is misrepresented. Request atomic writes to preserve the repository's boolean contract:

```suggestion
                .addQueryParameter("db", configuration.getDatabaseName()).addQueryParameter("precision", "ms")
                .addQueryParameter("accept_partial", "false").build();
```

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        point.getTags().entrySet().stream().sorted(Map.Entry.comparingByKey()).forEach(tag -> line.append(',')
+                .append(escapeTagKeyOrValue(tag.getKey())).append('=').append(escapeTagKeyOrValue(tag.getValue())));
+        line.append(' ').append(FIELD_VALUE_NAME).append('=').append(formattedValue.get()).append(' ')
+                .append(point.getTime().toEpochMilli());
+        return Optional.of(line.toString());
+    }
+
+    private Optional<String> formatFieldValue(@Nullable Object value) {
+        if (value instanceof String string) {
+            return Optional.of('"' + string.replace("\\", "\\\\").replace("\"", "\\\"") + '"');
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:190]: InfluxDB line protocol does not support newline characters in field or tag values. This escaping leaves `\n` untouched, so a multiline `StringType` state is split into extra protocol lines and the write fails or partially succeeds. Reject or otherwise handle unsupported newlines before building the batch, including measurement and tag inputs, and cover this serializer with unit tests.

Comment (lsiepel) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:190]: i cannot assis this finding without needed influx knowledge, but leave it here for your judgement

**Review by lsiepel — COMMENTED** [review]
Looked at all files and together with codex it seems there are some issues to address. 

_This is an initial AI-assisted review._

**Comment by maniac103** [comment]
> Wouldn;t this be a good oppertunity to remove V1 or call it deprecated and mark it for removal in openHAB 6

I share @cedricboon 's arguments here: InfluxDB 1, 2 and 3 are not direct successors to each other and thus can not be replaced by each other directly. I therefore don't think we should deprecate v1, at least not as long as they continue providing updates for v1 (which they do).

**Review by lsiepel — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBVersion.java (hunk 1 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBVersion.java:26]: Added the option and help text to config.xml and influxdb.properties

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBConfiguration.java (hunk 2 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBConfiguration.java:86]: Added a hasValidRetentionPolicy variable that takes V3 into account to evaluate the validity of the retention policy

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java (hunk 3 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java]: Ok, modified

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java (hunk 4 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java]: Added a top-level try-catch block

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java (hunk 5 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java]: Ok, modified

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java (hunk 6 above)
Comment (cedricboon) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java:211]: Ok, malformed input with newline characters are now discarded with a warning

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    @Override
+    public boolean checkConnectionStatus() {
+        final OkHttpClient currentClient = client;
+        if (currentClient == null) {
+            logger.warn("checkConnection: database is not connected");
+            return false;
+        }
+        try {
+            HttpUrl url = HttpUrl.get(configuration.getUrl()).newBuilder().addPathSegment("ping").build();
+            Request request = new Request.Builder().url(url).get().build();
```
Comment (lsiepel) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java]: InfluxDB 3 requires authentication on /ping by default, but this request omits the configured bearer token. A healthy authenticated instance therefore returns 401, so connect() reports failure and logs that the database is not ready even though the authenticated write/query requests can work. Add the same header used by those requests:

```suggestion
            Request request = new Request.Builder().url(url)
                    .header("Authorization", "Bearer " + configuration.getToken()).get().build();
```

Comment (lsiepel) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx3/InfluxDB3RepositoryImpl.java]: Confirmed by : https://docs.influxdata.com/influxdb3/core/api/server-information/

**Review by lsiepel — CHANGES_REQUESTED** [review]
Just one comment left, otherwise LGTM

**Comment by cedricboon** [comment]
Ok, last comment modified as well.
Thanks for the review!

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM
